crates/oxc_linter/src/rules/unicorn/prefer_default_parameters.rs

```diff
@@ -0,0 +1,535 @@
+use oxc_ast::{
+    AstKind,
+    ast::{
+        AssignmentOperator, AssignmentTarget, BindingPatternKind, Expression, FormalParameter,
+        LogicalOperator,
+    },
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{GetSpan, Span};
+use oxc_syntax::node::NodeId;
+
+use crate::{AstNode, context::LintContext, rule::Rule};
+
+fn prefer_default_parameters_diagnostic(span: Span, name: &str) -> OxcDiagnostic {
+    OxcDiagnostic::warn(format!("Prefer default parameters over reassignment for '{name}'."))
+        .with_help("Replace the reassignment with a default parameter.")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct PreferDefaultParameters;
+
+// See <https://github.com/oxc-project/oxc/issues/6050> for documentation details.
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Instead of reassigning a function parameter, default parameters should be used. The `foo = foo || 123` statement evaluates to `123` when `foo` is falsy, possibly leading to confusing behavior, whereas default parameters only apply when passed an `undefined` value.
+    /// This rule only reports reassignments to literal values.
+    ///
+    /// You should disable this rule if you want your functions to deal with `null` and other falsy values the same way as `undefined`.
+    /// Default parameters are exclusively applied [when `undefined` is received.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters#passing_undefined_vs._other_falsy_values).
+    /// However, we recommend [moving away from `null`](https://github.com/sindresorhus/meta/discussions/7).
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Using default parameters makes it clear that a parameter has a default value, improving code readability and maintainability.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// function abc(foo) {
+    /// 	foo = foo || 'bar';
+    /// }
+    ///
+    /// function abc(foo) {
+    /// 	const bar = foo || 'bar';
+    /// }
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// function abc(foo = 'bar') {}
+    ///
+    /// function abc(bar = 'bar') {}
+    ///
+    /// function abc(foo) {
+    /// 	foo = foo || bar();
+    /// }
+    /// ```
+    PreferDefaultParameters,
+    unicorn,
+    style,
+    pending,
+);
+
+fn is_literal(expr: &Expression) -> bool {
+    matches!(
+        expr.without_parentheses(),
+        Expression::StringLiteral(_)
+            | Expression::NumericLiteral(_)
+            | Expression::BooleanLiteral(_)
+            | Expression::NullLiteral(_)
+            | Expression::BigIntLiteral(_)
+            | Expression::RegExpLiteral(_)
+            | Expression::TemplateLiteral(_)
+    )
+}
```

Comment:
this function is unneccessary, you should be able to just call `expr.get_inner_expression().is_literal()`